### PR TITLE
Support trainable_parameters_mask in MaxTextTrainingEngine

### PR DESCRIPTION
**Pull request by niting**
## Description

When `trainable_parameters_mask` is specified, `get_optimizer()` wraps
the optimizer with `optax.multi_transform` to map non-matching parameters
to `optax.set_to_zero()`. While this prevents updates to frozen weights,
autodiff still computes non-zero gradients for all parameters.

In `MaxTextTrainingEngine._update_kernel`, `grad_norm` and
`apply_gradient_clipping` were computed over the unmasked gradients
before the optimizer zeroed the updates. On models with frozen components
(such as freezing router gate kernels in MoE RL training), this caused:
1. `grad_norm` to be inflated by frozen weights.
2. `apply_gradient_clipping` to aggressively over-clip trainable weights
   based on the inflated global norm.
3. Spikes on frozen parameters to trigger false-positive step skips when
   `skip_step_on_spikes` is enabled.

This change:
- Masks out frozen parameter gradients in `_update_kernel` before
  computing `grad_norm` and applying gradient clipping.
- Caches `_freeze_mask_fn` and `_freeze_mask` across eager and compiled
  execution paths in `MaxTextTrainingEngine`.
- Extracts `apply_trainable_parameters_mask` in `optimizers.py` with
  diagnostic logging of frozen parameter counts and sample paths.
- Adds unit tests in `maxtext_engine_test.py` covering eager and
  compiled execution with `trainable_parameters_mask`.

### CODE REVIEW

File: src/maxtext/training_engine/maxtext_engine.py [hunk 1]
```diff
@@ -1230,6 +1235,14 @@ def _update_kernel(self, state_pure, accumulated_grads, accumulated_denominator,
           lambda g: jnp.where(has_weights, g / safe_denominator.astype(g.dtype), jnp.zeros_like(g)),
           accumulated_grads,
       )
+      freeze_mask = self._freeze_mask
+      if freeze_mask is None and self._freeze_mask_fn is not None:
+        freeze_mask = self._freeze_mask_fn(accumulated_grads)
```
[the remark this answers is not in the document]
Comment (niting) [reply src/maxtext/training_engine/maxtext_engine.py:1241]: Fixed. Cached self._freeze_mask = freeze_mask in eager mode.

File: src/maxtext/training_engine/maxtext_engine.py [hunk 2]
```diff
@@ -1230,6 +1235,14 @@ def _update_kernel(self, state_pure, accumulated_grads, accumulated_denominator,
...
+      if freeze_mask is not None:
+        def _apply_freeze(g, m):
+          is_frozen = m.get_value() if hasattr(m, "get_value") else m
+          return jnp.where(is_frozen, jnp.zeros_like(g), g)
```
[the remark this answers is not in the document]
Comment (niting) [reply src/maxtext/training_engine/maxtext_engine.py:1245]: Fixed. Switched to Python if/else for static trace-time branch elimination.

**Comment by niting** [comment]
@gemini-code-assist review
